modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSqlMultiStatementTxTest.java

```diff
@@ -0,0 +1,219 @@
+    private int numberOfFinishedTransactionsOnStart;
+
+    @BeforeAll
+    void createTable() {
+        sql("CREATE TABLE test (id INT PRIMARY KEY)");
+    }
+
+    @AfterEach
+    void clearTable() {
+        sql("DELETE FROM test");
+    }
+
+    @BeforeEach
+    void saveTxCount() {
+        numberOfFinishedTransactionsOnStart = txManager().finished();
+    }
+
+    void checkFinished(int expected) {
+        assertEquals(numberOfFinishedTransactionsOnStart + expected, txManager().finished());
+    }
+
+    @Test
+    void emptyTransactionControlStatement() {
+        List<AsyncSqlCursor<List<Object>>> cursors = fetchAllCursors(
+                runScript("START TRANSACTION; COMMIT"));
+
+        assertThat(cursors, hasSize(2));
+
+        validateSingleResult(cursors.get(0));
+        validateSingleResult(cursors.get(1));
+
+        checkFinished(1);
+    }
+
+    @Test
+    void basicTxStatements() {
+        fetchAllCursors(runScript(
+                "START TRANSACTION;"
+                        + "INSERT INTO test VALUES(0);"
+                        + "INSERT INTO test VALUES(1);"
+                        + "COMMIT;"
+
+                        + "START TRANSACTION;"
+                        + "COMMIT;"
+
+                        + "START TRANSACTION;"
+                        + "INSERT INTO test VALUES(2);"
+                        + "COMMIT;"
+        ));
+
+        checkFinished(3);
+
+        assertQuery("select count(id) from test")
+                .returns(3L).check();
+    }
+
+    @Test
+    void scriptTransactionRollsBackImplicitly() {
+        {
+            fetchAllCursors(runScript("START TRANSACTION;"));
+            checkNoPendingTransactions();
+            checkFinished(1);
+        }
+
+        {
+            fetchAllCursors(runScript("START TRANSACTION;"
+                    + "INSERT INTO test VALUES(0);"
+                    + "INSERT INTO test VALUES(1);"
+                    + "COMMIT;"
+                    + "START TRANSACTION;"
+                    + "INSERT INTO test VALUES(2);"));
+
+            checkFinished(3);
+
+            checkNoPendingTransactions();
+
+            assertQuery("select count(id) from test")
+                    .returns(2L).check();
+        }
+    }
+
+    @Test
+    void scriptTransactionRollsBackOnError() {
+        {
+            AsyncSqlCursor<List<Object>> cursor = runScript(
+                    "START TRANSACTION;"
+                    + "INSERT INTO test VALUES(2);"
+                    + "INSERT INTO test VALUES(2/0);"
+                    + "COMMIT;"
+            );
+
+            assertThrowsSqlException(RUNTIME_ERR, "/ by zero", () -> fetchAllCursors(cursor));
+
+            checkNoPendingTransactions();
+            checkFinished(1);
+
+            assertQuery("select count(id) from test")
+                    .returns(0L).check();
+        }
+    }
+
+    @Test
+    void commitWithoutOpenTransactionDoesNothing() {
```

Comment:
Should DDL fail transaction in the next case?
```
START TRANSACTION;
INSERT INTO test VALUES (0);
CREATE TABLE ....;
COMMIT;
```

Reply:
yes

Reply:
`ddlInsideExplicitTransactionFails` covers this case